[ES|QL] Automatically encapsulate index names with special chars with quotes

qn895 opened the pull request:
## Summary

Addresses https://github.com/elastic/kibana/issues/186411. This PR introduces quoted index names when auto populated/suggested if the index name contains special characters.

https://github.com/elastic/kibana/assets/43350163/94a37a82-9921-4419-b7d4-c8c50cc509f3

If user is already prefixing index name with a quote, e.g. `from "a` it will suggest as normal.


### Checklis

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.15"]} ONMERGE-->

qn895 commented:
/ci

qn895 commented:
/ci

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/220460)
* Commit: 5cc965a94313decb238327c360057da0c023a113

### Failed CI Steps


### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/5bc0c7865e11b9b699d282c488286779b344f0e6) | [after](https://github.com/elastic/kibana/commit/5cc965a94313decb238327c360057da0c023a113) | diff |
| --- | --- | --- | --- |
| `kbnUiSharedDeps-srcJs` | 3.2MB | 3.2MB | +285.0B |



### History
* :broken_heart: [Build #220437](https://buildkite.com/elastic/kibana-pull-request/builds/220437) failed bf2715c3f10aa84db0c46299f539acc50b230ea0

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220460","number":220460,"commit":"5cc965a94313decb238327c360057da0c023a113"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220437","number":220437,"commit":"bf2715c3f10aa84db0c46299f539acc50b230ea0"}],"number":220460}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/kibana-esql (Team:ESQL)

drewdaemon reviewed (commented):
This is definitely on the right track. At first, your approach of quoting by default and only unquoting if we're sure it doesn't need it struck me as odd. However, it grew on me because there are two mistakes we can potentially make
1. we quote an index that doesn't need it
2. we unquote an index that does need it

Number 1 is basically fine, but number 2 actually makes the query fail, so I prefer to err on the side of the first mistake.

On that note, have we tested this with integrations? As far as I can tell, we are quoting this index unnecessarily:

https://github.com/elastic/kibana/assets/315764/0d0dc24c-3566-4ba0-9a4b-0694d852862e

Final thought: you can now use triple-quotes `"""` to escape things. Let's add these triples to [these monaco settings](https://github.com/elastic/kibana/blob/main/packages/kbn-monaco/src/esql/language.ts#L38-L48) so that Monaco can help us out here:

https://github.com/elastic/kibana/assets/315764/e71335b4-9407-4e89-804a-c02c5bd5c18d

qn895 commented:
Thanks for the hint about the triple quotes. [I added support here](https://github.com/elastic/kibana/pull/187899/commits/9ab8f4a260b23310ed8a6b220a979d8aa1f76c26) for triple quote auto closing and surrounding pairs. I also updated the regex to reflect support for things that can be unquoted (based on `fragment UNQUOTED_SOURCE_PART
    : ~[:"=|,[\]/ \t\r\n]
    | '/' ~[*/]`)

drewdaemon commented on packages/kbn-esql-validation-autocomplete/src/shared/helpers.ts:
```diff
@@ -566,6 +566,10 @@ export function isRestartingExpression(text: string) {
   return getLastCharFromTrimmed(text) === ',';
 }
 
+export function shouldBeQuotedSource(text: string) {
```
Are we testing this anywhere?

qn895 replied:
Updated here [eccfb5e](https://github.com/elastic/kibana/pull/187899/commits/eccfb5e1de1cfd57681335ddf5ed806ed13fddc3)

drewdaemon commented on packages/kbn-esql-validation-autocomplete/src/autocomplete/autocomplete.ts:
```diff
@@ -857,19 +858,28 @@ async function getExpressionSuggestionsByType(
         suggestions.push(...(policies.length ? policies : [buildNoPoliciesAvailableDefinition()]));
       } else {
         const index = getSourcesFromCommands(commands, 'index');
+        const shouldRemoveQuote = isNewExpression && innerText.includes('"');
```
```suggestion
        const canRemoveQuote = isNewExpression && innerText.includes('"');
```

qn895 replied:
Updated here [eccfb5e](https://github.com/elastic/kibana/pull/187899/commits/eccfb5e1de1cfd57681335ddf5ed806ed13fddc3)

drewdaemon reviewed (approved):
Thank you!

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/220738)
* Commit: 9ab8f4a260b23310ed8a6b220a979d8aa1f76c26

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/5bc0c7865e11b9b699d282c488286779b344f0e6) | [after](https://github.com/elastic/kibana/commit/9ab8f4a260b23310ed8a6b220a979d8aa1f76c26) | diff |
| --- | --- | --- | --- |
| `kbnUiSharedDeps-srcJs` | 3.2MB | 3.2MB | +345.0B |



### History
* :yellow_heart: [Build #220460](https://buildkite.com/elastic/kibana-pull-request/builds/220460) was flaky 5cc965a94313decb238327c360057da0c023a113
* :broken_heart: [Build #220437](https://buildkite.com/elastic/kibana-pull-request/builds/220437) failed bf2715c3f10aa84db0c46299f539acc50b230ea0

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220738","number":220738,"commit":"9ab8f4a260b23310ed8a6b220a979d8aa1f76c26"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220460","number":220460,"commit":"5cc965a94313decb238327c360057da0c023a113"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220437","number":220437,"commit":"bf2715c3f10aa84db0c46299f539acc50b230ea0"}],"number":220738}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/220766)
* Commit: 76cdad1f532de0b5cd0a43d3ebdfbe960586a9cf

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/5bc0c7865e11b9b699d282c488286779b344f0e6) | [after](https://github.com/elastic/kibana/commit/76cdad1f532de0b5cd0a43d3ebdfbe960586a9cf) | diff |
| --- | --- | --- | --- |
| `kbnUiSharedDeps-srcJs` | 3.2MB | 3.2MB | +345.0B |



### History
* :green_heart: [Build #220738](https://buildkite.com/elastic/kibana-pull-request/builds/220738) succeeded 9ab8f4a260b23310ed8a6b220a979d8aa1f76c26
* :yellow_heart: [Build #220460](https://buildkite.com/elastic/kibana-pull-request/builds/220460) was flaky 5cc965a94313decb238327c360057da0c023a113
* :broken_heart: [Build #220437](https://buildkite.com/elastic/kibana-pull-request/builds/220437) failed bf2715c3f10aa84db0c46299f539acc50b230ea0

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220766","number":220766,"commit":"76cdad1f532de0b5cd0a43d3ebdfbe960586a9cf"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220738","number":220738,"commit":"9ab8f4a260b23310ed8a6b220a979d8aa1f76c26"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220460","number":220460,"commit":"5cc965a94313decb238327c360057da0c023a113"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/220437","number":220437,"commit":"bf2715c3f10aa84db0c46299f539acc50b230ea0"}],"number":220766}
buildkite-pr-comment-->

stratoula commented:
Merging as I want to backport in 8.15

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/188052">](https://github.com/elastic/kibana/pull/188052)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
